src/bin/cargo/cli.rs

```diff
@@ -9,7 +9,19 @@ use super::commands;
 use command_prelude::*;
 
 pub fn main(config: &mut Config) -> CliResult {
-    let args = cli().get_matches_safe()?;
+    let args = match cli().get_matches_safe() {
+        Ok(args) => args,
+        Err(e) => {
+            if e.kind == clap::ErrorKind::UnrecognizedSubcommand {
+                // An unrecognized subcommand might be an external subcommand.
+                let cmd = &e.info.as_ref().unwrap()[0].to_owned();
```

Comment:
Reading clap's docs, it is not clear that this wont panic on unwrap or `[0]` in some cases. But let's leave this as is: this is an error-path anyway, so loud panic should be not cause problems, and would be easier to fix than slient do-nothing.